### ISSUE TITLE
Neo4j 3.5.1->4 (embedded): Wrong relationships when traversing PathFinder result & Node.getRelationships(Direction, RelationshipType) 

### ISSUE THREAD

**Issue by efruttero** [neo4j/neo4j#12155]
**Neo4j Version:** Community-Edition, 3.5.1 and 3.5.2  
**Operating System:** Windows 10 Pro
**API:** Java API

Hello,

I have a problem that appeared in 3.5.1 and 3.5.2.
3.4.12 and 3.5.0 are OK.

I can't share my dataset, but I am **very willing** to assist you asserting/probing the problem.
I have built the same dataset, for each Neo4j version.
The following script runs fine on 3.4.12 and 3.5.0 (i.e, **badRels** is empty)
All the **check(...)** are OK for all versions.

Thanks again !  

Eric Fruttero

Model:  
(Segment) -[:SEGMENT_NEIGHBOR]-> (Segment)
(Segment) -[:SEGMENT__CLOSEST_BOOTH]->(Booth)

```kotlin
val Segment = Label.label("Segment")
val Booth = Label.label("Booth")
val SEGMENT_NEIGHBOR = RelationshipType.withName("SEGMENT_NEIGHBOR")
val SEGMENT__CLOSEST_BOOTH = RelationshipType.withName("SEGMENT__CLOSEST_BOOTH")

val Node.isSegment: Boolean get() = this.hasLabel(Segment)
val Node.isBooth: Boolean get() = this.hasLabel(Booth)

val segmentsPathFinder = GraphAlgoFactory.shortestPath(
                             PathExpanders.forTypeAndDirection<Node>(
                                 SEGMENT_NEIGHBOR,
                                 Direction.BOTH),
                             2)

val segments = db.findNodes(Segment).asSequence().toSet()
check(segments.size == 4261)  // same (4) datasets 3.4.12 and 3.5.X

// 1 - for each segment, get booths, OK
segments.forEach { segment ->
    check(segment.isSegment)
    val relsToBooths = segment.getRelationships(Direction.OUTGOING, SEGMENT__CLOSEST_BOOTH)
    check(relsToBooths.all { it.isType(SEGMENT__CLOSEST_BOOTH)})
    check(relsToBooths.all { it.startNode == segment})
    check(relsToBooths.all { it.endNode.isBooth})
}

val badSegments = HashSet<Node>()
val badRels = HashSet<Relationship>()

// 2 - Finding paths through segments, for each segment-in-path, get booths, sometimes KO for 3.5.1/2
var pathsCount = 0
segments.forEach {segment1 ->
    segments.forEach { segment2 ->
        val path = segmentsPathFinder.findSinglePath(segment1, segment2)
        if (path == null)
        {
            return@forEach
        }

        pathsCount++
        path.nodes().forEach {segmentInPath ->
            check(segmentInPath.isSegment)
            val relsToBooths = segmentInPath.getRelationships(Direction.OUTGOING, SEGMENT__CLOSEST_BOOTH)
            relsToBooths.forEach { rel ->
                if (rel.isType(SEGMENT__CLOSEST_BOOTH))  // OK
                {
                    check(rel.startNode == segmentInPath)
                    check(rel.endNode.isBooth)
                }
                else  // WRONG RELATIONSHIP TYPE (3.5.1/2), always SEGMENT_NEIGHBOR
                {
                    check(rel.isType(SEGMENT_NEIGHBOR))
                    check(rel.startNode.isSegment) // rel.startNode SOMETIMES != segmentInPath
                    badSegments.add(segmentInPath)
                    badSegments.add(rel.startNode)
                    badRels.add(rel)
                }
            }
        }
    }
}

check(pathsCount == 31375) // same (4) datasets 3.4.12 and 3.5.X

// 3 - Some info on badRels
if (badRels.isNotEmpty())  // 3.5.1/2
{
    check(badRels.size == 385) // same (4) datasets 3.4.12 and 3.5.X
    val relTypes = badRels.map {it.type}.toSet()
    val startLabels = badRels.flatMap { it.startNode.labels }.toSet()
    val endLabels = badRels.flatMap { it.endNode.labels }.toSet()

    check(relTypes == setOf(SEGMENT_NEIGHBOR))
    check(startLabels == setOf(Segment))
    check(endLabels == setOf(Segment))
}

// 4 - all badSegments now out of PathFinder are OK vs booths
badSegments.forEach { segment ->
    check(segment.isSegment)
    val relsToBooths = segment.getRelationships(Direction.OUTGOING, SEGMENT__CLOSEST_BOOTH)
    check(relsToBooths.all { it.isType(SEGMENT__CLOSEST_BOOTH)})
    check(relsToBooths.all { it.startNode == segment})
    check(relsToBooths.all { it.endNode.isBooth})
}

// 5 - all original segments still OK vs booths
segments.forEach { segment ->
    check(segment.isSegment)
    val relsToBooths = segment.getRelationships(Direction.OUTGOING, SEGMENT__CLOSEST_BOOTH)
    check(relsToBooths.all { it.isType(SEGMENT__CLOSEST_BOOTH)})
    check(relsToBooths.all { it.startNode == segment})
    check(relsToBooths.all { it.endNode.isBooth})
}
```

**Comment by efruttero** [comment]
I created a minimal 3.5.3 obfuscated and without properties version (besides the Segment/Booth part) of my dataset. It's 30 MB zipped.
Let me know if you're interested to get it and how to proceed.

Thanks

**Comment by efruttero** [comment]
Hello,

I would have thought that this bug is in the "**_Ho ? Interesting.._**" category.
Looking quite core/essential to me. :)

**Comment by efruttero** [comment]
Still there in 3.5.4

**Comment by efruttero** [comment]
Fixed by 3.5.13.

Thanks ! :)
